openupgrade_scripts/scripts/base/14.0.1.3/end-migration.py

```diff
@@ -3,6 +3,20 @@
 from openupgradelib import openupgrade
 
 
+def uninstall_conflicting_it_edi(cr):
+    it_edi_conflicting_modules = ("l10n_it_edi", "l10n_it_fatturapa")
+    if all(
+        openupgrade.is_module_installed(cr.cr, m) for m in it_edi_conflicting_modules
+    ):
+        it_edi_module = cr["ir.module.module"].search(
+            [
+                ("name", "=", "l10n_it_edi"),
+            ],
+            limit=1,
+        )
+        it_edi_module.button_uninstall()
```

Comment:
Why doing 2 times the same?

Reply:
Thanks for having a look!
> Mark as 'to_remove' to avoid raising a conflict; it will be installed anyway,
but we will uninstall it for good in end-migration.

I wrote it in a comment in the code in https://github.com/OCA/OpenUpgrade/pull/3595/files#diff-aec3724a38e5c01c0ea5b62012c57a9145be5ab7058f222be86607cb7c50301fR93-R94.

Reply:
Yeah, but why it will be installed? You are saying in other comment that is not auto-install. And if it's installed by another module, uninstalling this one will uninstall the other module as well.

Reply:
Just summarizing a bit and adding a few more details:
In `13.0` DB I have:

| name | state |
|-|-|
 l10n_it_edi | installed
 l10n_it_fatturapa | to upgrade

When I upgrade to `14.0` without this PR, the error in the description is raised:
> odoo.exceptions.UserError: I moduli "ITA - Fattura elettronica - Base" e "Italy - E-invoicing" non sono compatibili.

If I include only the `pre-migrate` part of this PR, the migration concludes successfully BUT `l10n_it_edi` and `l10n_it_fatturapa` are both installed; that is why I also added the `end-migration` to uninstall the module.

---
Now back to your question:
> Yeah, but why it will be installed?

`l10n_it_edi` is added to the graph of modules to be loaded in https://github.com/odoo/odoo/blob/a46d20c89c24a4e60faef09cf8ad7721bd29d6cc/odoo/modules/loading.py#L341-L345:
```
        cr.execute("SELECT name from ir_module_module WHERE state IN %s" ,(tuple(states),))
        module_list = [name for (name,) in cr.fetchall() if name not in graph]
        if not module_list:
            break
        graph.add_modules(cr, module_list, force)
```
because it is called including the state `to_remove` in https://github.com/odoo/odoo/blob/a46d20c89c24a4e60faef09cf8ad7721bd29d6cc/odoo/modules/loading.py#L455-L457:
```
            processed_modules += load_marked_modules(cr, graph,
                ['installed', 'to upgrade', 'to remove'],
                force, status, report, loaded_modules, update_module, models_to_check)
```

Odoo thinks that `l10n_it_edi` has to be updated because the following condition is True:
```
        needs_update = (
            hasattr(package, "init")
            or hasattr(package, "update")
            or package.state in ("to install", "to upgrade")
        )
```
(from https://github.com/odoo/odoo/blob/a46d20c89c24a4e60faef09cf8ad7721bd29d6cc/odoo/modules/loading.py#L161-L165)
more precisely, `hasattr(package, "update")` is True because it is assigned in https://github.com/odoo/odoo/blob/cbab786eb2e153eb7eefbd994cb261fa96e56800/odoo/modules/graph.py#L85-L87:
```
                for kind in ('init', 'demo', 'update'):
                    if package in tools.config[kind] or 'all' in tools.config[kind] or kind in force:
                        setattr(node, kind, True)
```
More precisely, `package in tools.config[kind]` is True.

Reply:
Thank you for the detailed explanation. Wow, such a difficult thing to handle...